python/ray/util/sgd/v2/worker_group.py

```diff
@@ -1,32 +1,137 @@
-from typing import Callable, Any, List
+from typing import Callable, List, TypeVar
+
+from inspect import signature
 
 import ray
 from ray.types import ObjectRef
 
+T = TypeVar("T")
+
 
 class WorkerGroup:
+    """
+    A group of workers (Ray Actors) that can execute arbitrary functions.
+
+    ``WorkerGroup`` launches Ray actors according to the given
+    specification. It can then execute arbitrary Python functions in each of
+    these workers.
+
+    If not enough resources are available to launch the actors, the Ray
+    cluster will automatically scale up if autoscaling is enabled.
+
+    Args:
+        num_workers (int): The number of workers (Ray actors) to launch.
+            Defaults to 1.
+        num_cpus_per_worker (float): The number of CPUs to reserve for each
+            worker. Fractional values are allowed. Defaults to 1.
+        num_gpus_per_worker (float): The number of GPUs to reserve for each
+            worker. Fractional values are allowed. Defaults to 0.
+
+    Example:
+
+        .. code_block:: python
+
+        worker_group = WorkerGroup(num_workers=2)
+        output = worker_group.execute(lambda: 1)
+        assert len(output) == 2
+        assert all(o == 1 for o in output)
+    """
+
     def __init__(self,
                  num_workers: int = 1,
-                 num_cpus_per_worker: int = 1,
-                 num_gpus_per_worker: int = 0):
+                 num_cpus_per_worker: float = 1,
+                 num_gpus_per_worker: float = 0):
+
+        if num_workers <= 0:
+            raise ValueError("The provided `num_workers` must be greater "
+                             f"than 0. Received num_workers={num_workers} "
+                             f"instead.")
+        if num_cpus_per_worker < 0 or num_gpus_per_worker < 0:
+            raise ValueError("The number of CPUs and GPUs per worker must "
+                             "not be negative. Received "
+                             f"num_cpus_per_worker={num_cpus_per_worker} and "
+                             f"num_gpus_per_worker={num_gpus_per_worker}.")
+
+        self.num_workers = num_workers
+        self.num_cpus_per_worker = num_cpus_per_worker
+        self.num_gpus_per_worker = num_gpus_per_worker
         self.workers = self._start_workers()
 
     def _start_workers(self):
-        pass
+        remote_cls = ray.remote(
+            num_cpus=self.num_cpus_per_worker,
+            num_gpus=self.num_gpus_per_worker)(BaseWorker)
+        return [remote_cls.remote() for _ in range(self.num_workers)]
+
+    def restart(self):
+        """Restarts all the workers in this worker group."""
+        self.workers = self._start_workers()
```

Comment:
Can we just merge this with `_start_workers` into a public method that begins with an assertion that `self.workers` is empty? Otherwise we run into the problem where the workers might not be able to gracefully shut down.

Reply:
Actually do we want the assertion? `restart` should restart the workers regardless of their current state (whether they are alive or not).

Reply:
Oh I see what you mean. Couldn't we just call `shutdown` in `restart`?

Reply:
Actually nvm, your suggestions sounds good.